ethereum/core/src/main/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProvider.java

```diff
@@ -92,4 +109,81 @@ public WorldStateProofProvider(final WorldStateStorage worldStateStorage) {
         b -> b,
         b -> b);
   }
+
+  public boolean isValidRangeProof(
+      final Bytes32 startKeyHash,
+      final Bytes32 endKeyHash,
+      final Bytes32 rootHash,
+      final List<Bytes> proofs,
+      final TreeMap<Bytes32, Bytes> keys) {
+
+    // check if it's monotonic increasing
+    if (!Ordering.natural().isOrdered(keys.keySet())) {
+      return false;
+    }
+
+    // when proof is empty we need to have all the keys to reconstruct the trie
+    if (proofs.isEmpty()) {
+      final MerklePatriciaTrie<Bytes, Bytes> trie =
+          new SimpleMerklePatriciaTrie<>(Function.identity());
+      // add the received keys in the trie
+      for (Map.Entry<Bytes32, Bytes> key : keys.entrySet()) {
+        trie.put(key.getKey(), key.getValue());
+      }
+
+      return rootHash.equals(trie.getRootHash());
+    }
+
+    // reconstruct a part of the trie with the proof
+    Map<Bytes32, Bytes> proofsEntries = Collections.synchronizedMap(new HashMap<>());
+    for (Bytes proof : proofs) {
+      proofsEntries.put(Hash.hash(proof), proof);
+    }
+
+    if (keys.isEmpty()) {
+      final MerklePatriciaTrie<Bytes, Bytes> trie =
+          new StoredMerklePatriciaTrie<>(
+              new InnerNodeDiscoveryManager<>(
+                  (location, hash) -> Optional.ofNullable(proofsEntries.get(hash)),
+                  Function.identity(),
+                  Function.identity(),
+                  startKeyHash,
+                  endKeyHash,
+                  false),
+              rootHash);
+      try {
+        // check is there is not missing element
+        trie.entriesFrom(startKeyHash, Integer.MAX_VALUE);
+      } catch (MerkleTrieException e) {
+        return false;
+      }
+      return true;
+    }
+
+    // search inner nodes in the range created by the proofs and remove
+    final InnerNodeDiscoveryManager<Bytes> snapStoredNodeFactory =
+        new InnerNodeDiscoveryManager<>(
+            (location, hash) -> Optional.ofNullable(proofsEntries.get(hash)),
+            Function.identity(),
+            Function.identity(),
+            startKeyHash,
+            keys.lastKey(),
+            true);
+    final MerklePatriciaTrie<Bytes, Bytes> trie =
+        new StoredMerklePatriciaTrie<>(snapStoredNodeFactory, rootHash);
+    trie.visitAll(node -> {});
```

Comment:
what is this doing?  Just ensuring the trie is well formed?

edit: ah, looks like it is filling out innerNodes of the InnerNodeDiscoveryManager.  A comment might help to explain this side effect